src/common/include/sirf/common/num_ref.h

```diff
@@ -0,0 +1,323 @@
+#pragma once
+
+#include <complex>
+#include <typeinfo>
+
+#include <ismrmrd/ismrmrd.h>
+
+namespace sirf {
+
+	class aNumRef {
+	public:
+		virtual complex_double_t complex_double() const = 0;
+		virtual complex_float_t complex_float() const = 0;
+		virtual operator float() const = 0;
+		virtual void assign(const aNumRef& ref) = 0;
+		aNumRef& operator=(const aNumRef& ref)
+		{
+			assign(ref);
+			return *this;
+		}
+		virtual void set_ptr(void* ptr) = 0;
+		virtual void copy(const aNumRef& ref) = 0;
+	};
+
+	class FloatRef : public aNumRef {
+	public:
+		FloatRef(float* ptr = 0, int dummy = 0) : ptr_(ptr)
+		{}
+		virtual complex_double_t complex_double() const
+		{
+			return complex_double_t(*ptr_);
+		}
+		virtual complex_float_t complex_float() const
+		{
+			return complex_float_t(*ptr_);
+		}
+		virtual operator float() const
+		{
+			return *ptr_;
+		}
+		template<typename T>
+		FloatRef& operator=(T v)
+		{
+			*ptr_ = v;
+			return *this;
+		}
+		virtual void assign(const aNumRef& a_ref)
+		{
+			const FloatRef& ref = (const FloatRef&)a_ref;
+			*ptr_ = float(ref);
+		}
+		aNumRef& operator=(const aNumRef& ref)
+		{
+			assign(ref);
+			return *this;
+		}
+		void set_ptr(void* ptr)
+		{
+			ptr_ = (float*)ptr;
+		}
+		void copy(const aNumRef& a_ref)
+		{
+			const FloatRef& ref = (const FloatRef&)a_ref;
+			ptr_ = ref.ptr_;
+		}
+	private:
+		float* ptr_;
+	};
+
+	template <typename Type>
+	ISMRMRD::ISMRMRD_DataTypes TypeID(Type t)
+	{
+		if (typeid(Type) == typeid(complex_double_t))
+			return ISMRMRD::ISMRMRD_CXDOUBLE;
+		else if (typeid(Type) == typeid(complex_float_t))
+			return ISMRMRD::ISMRMRD_CXFLOAT;
+		else if (typeid(Type) == typeid(double))
+			return ISMRMRD::ISMRMRD_DOUBLE;
+		else if (typeid(Type) == typeid(float))
+			return ISMRMRD::ISMRMRD_FLOAT;
+		else if (typeid(Type) == typeid(int))
+			return ISMRMRD::ISMRMRD_INT;
+		else if (typeid(Type) == typeid(unsigned int))
+			return ISMRMRD::ISMRMRD_UINT;
+		else if (typeid(Type) == typeid(short))
+			return ISMRMRD::ISMRMRD_SHORT;
+		else if (typeid(Type) == typeid(unsigned short))
+			return ISMRMRD::ISMRMRD_USHORT;
+		else
+			throw std::invalid_argument
+			(std::string("unsupported numeric type ") + typeid(Type).name());
+	}
+
+	class NumRef : public aNumRef {
+	public:
+		NumRef(void* ptr = 0, int type = (int)ISMRMRD::ISMRMRD_FLOAT) :
+			ptr_(ptr), abs_(true), type_(type)
+		{}
+		NumRef(const NumRef& ref) :
+			ptr_(ref.ptr_), abs_(ref.abs_), type_(ref.type_)
+		{}
+		void set_complex_to_real_mode(char m)
+		{
+			abs_ = (m == 'a');
+		}
+		char type() const
+		{
+			return type_;
+		}
+		virtual void set_ptr(void* ptr)
+		{
+			//std::cout << (size_t)ptr << '\n';
+			ptr_ = ptr;
+		}
+		virtual void copy(const aNumRef& a_ref)
+		{
+			const NumRef& ref = (const NumRef&)a_ref;
```

Comment:
why do we now that `a_ref` is a `NumRef`? Would need to use `dynamic_cast` to be sure

Reply:
actually should have been

    void copy(const NumRef& ref)

and no cast; will do after merging new `num_ref.h` by Richard